core/binding/src/store/map.rs

```diff
@@ -1 +1,163 @@
+
+        let inner_k: Result<Vec<K>, _> = inner_u8
+            .into_iter()
+            .map(|v| <_>::decode_fixed(Bytes::from(v)))
+            .collect();
+
+        let inner = inner_k.map_err(|_| rlp::DecoderError::Custom("decode K from bytes fail"))?;
+
+        Ok(FixedKeys { inner })
+    }
+}
+
+impl<K: FixedCodec> FixedCodec for FixedKeys<K> {
+    fn encode_fixed(&self) -> ProtocolResult<Bytes> {
+        Ok(Bytes::from(rlp::encode(self)))
+    }
+
+    fn decode_fixed(bytes: Bytes) -> ProtocolResult<Self> {
+        Ok(rlp::decode(bytes.as_ref()).map_err(FixedCodecError::from)?)
+    }
+}
+
+pub struct DefaultStoreMap<S: ServiceState, K: FixedCodec + PartialEq, V: FixedCodec> {
+    state:    Rc<RefCell<S>>,
+    var_name: Hash,
+    keys:     FixedKeys<K>,
+    phantom:  PhantomData<V>,
+}
+
+impl<S: ServiceState, K: FixedCodec + PartialEq, V: FixedCodec> DefaultStoreMap<S, K, V> {
+    pub fn new(state: Rc<RefCell<S>>, name: &str) -> Self {
+        let var_name = Hash::digest(Bytes::from(name.to_owned() + "map"));
+
+        let opt_bs: Option<Bytes> = state
+            .borrow()
+            .get(&var_name)
+            .expect("get map should not fail");
+
+        let keys = if let Some(bs) = opt_bs {
+            <_>::decode_fixed(bs).expect("decode keys should not fail")
+        } else {
+            FixedKeys { inner: Vec::new() }
+        };
+
+        Self {
+            state,
+            var_name,
+            keys,
+            phantom: PhantomData,
+        }
+    }
+
+    fn get_map_key(&self, key: &K) -> ProtocolResult<Hash> {
+        let mut name_bytes = self.var_name.as_bytes().to_vec();
+        name_bytes.extend_from_slice(key.encode_fixed()?.as_ref());
+
+        Ok(Hash::digest(Bytes::from(name_bytes)))
+    }
+}
+
+impl<S: ServiceState, K: FixedCodec + PartialEq, V: FixedCodec> StoreMap<K, V>
+    for DefaultStoreMap<S, K, V>
+{
+    fn get(&self, key: &K) -> ProtocolResult<V> {
+        if self.keys.inner.contains(key) {
+            let mk = self.get_map_key(key)?;
+            self.state.borrow().get(&mk)?.map_or_else(
+                || <_>::decode_fixed(Bytes::new()).map_err(|_| StoreError::DecodeError.into()),
+                |v| Ok(v),
+            )
+        } else {
+            Err(StoreError::GetNone.into())
+        }
+    }
+
+    fn contains(&self, key: &K) -> ProtocolResult<bool> {
+        Ok(self.keys.inner.contains(key))
+    }
+
+    // TODO(@zhounan): Atomicity of insert(k, v) and insert self.keys to
+    // ServiceState is not guaranteed for now That must be settled soon after.
+    fn insert(&mut self, key: K, value: V) -> ProtocolResult<()> {
+        let mk = self.get_map_key(&key)?;
+        self.state.borrow_mut().insert(mk, value)?;
+
+        if !self.contains(&key)? {
+            self.keys.inner.push(key);
+            self.state
+                .borrow_mut()
+                .insert(self.var_name.clone(), self.keys.encode_fixed()?)
+        } else {
```

Comment:
Useless else.

I bet you didn't run `cargo clippy --all targets --all features`.